### DISCUSSION TITLE
Windows Batch - Only Modify Files Smaller Than a Specified Dimension

### DISCUSSION THREAD

**Discussion by flaccidbagel** [ImageMagick/ImageMagick#3025] [Help] [answered]
Good day.

I've been looking into ImageMagick to shrink down a collection of photos I've got from over the years. From testing, it works well using a simple batch script to just resize anything over a certain width. Essentially, the script just sets the top level folder and then runs a recursive for loop for pngs and jpgs.

```
for /r %%A in (*.png) do mogrify -resize 1920^> "%%A"
for /r %%A in (*.jpg) do mogrify -resize 1920^> "%%A"
```

Its quick and works well. However, one thing I've noticed is that this command will modify files every single time its ran, including if the file is the proper size. So if the image width is set to 1920, the resulting image keeps the current dimensions, but it still reprocesses the file, changing the file size slightly. I'm worried that this will eventually result in noticeable drops in quality as it keep processing the file each time its ran.

Is it possible to only run the resize command if the image actually doesn't meet the criteria set, to avoid constantly modifying the files? If there's a simple way to do so via batch I'm all ears.

**Comment by snibgo** [comment]
IM reads a file, does whatever processing you want (if any), and writes to a file. JPEG compression is lossy, so information is lost when the file is written. If you do this repeatedly, the result will be horrible.

IM does not have an operation that means "write the file if, and only if, the image has changed".

I suggest that you either:

(a) Arrange your workflow so you don't resize multiple times. For example, you might have a directory for incoming files, and a different directory for resized files.

(b) In your BAT script, test the size of the image in one magick command, then IF the width is greater than 1920 then run another magick command that does the resize. This would need a loop eg a "for" loop.

**Comment by GeeMack** [comment, marked as answer]
Here is a command using ImageMagick v7 in Windows syntax that uses a "for" loop to read in each JPG file in a directory and its sub-directories, resize every image that is wider than 1920 pixels, and write all those resized images to their original filenames. It will leave any image untouched if it's not wider than 1920.

    for /r %I in ( *.jpg ) do (
       magick "%I" -set filename:0 "%[d]\\%[f]" ^
       ^( xc: -set filename:0 bogus ^) ^
       -set option:switch "%[fx:u.w>1920]" ^
       -resize 1920^> -delete %[switch] "%[filename:0]"
    ) 2>NUL

It starts by reading in the image and making a single 1x1 pixel placeholder image so there are 2 images in the list.

Then it tests the input image width to create a switch to indicate which of the 2 images will get deleted before writing the output file.

Next it resizes the input if it's wider than 1920, and won't do anything if it's not. The placeholder image doesn't get resized of course because it's less than 1920 pixels.

Then the command deletes the placeholder if the image was resized, or deletes the image if it wasn't, and attempts to write the output image.

The trick here works when IM tries to write the output. It will succeed if the input was resized, but tries to write the "bogus" image if the input wasn't resized. Since that "bogus" image has no filename suffix IM can't write it because it doesn't know how. If that happens IM will generate an error message, but that message is sent to Windows' NUL with "2>NUL" at the end of the loop.

Double those percent signs "%%" if the command is in a BAT script.

EDITED: Modified it a bit to make the command work recursively.

EDITED: You should be able to run all the JPG and PNG files in a single "for" loop by putting all your wildcards inside the same parentheses like this...

    for /r %I in ( *.jpg *.png  ) do ( ... )

**Reply by flaccidbagel** [reply]
Thanks for the detailed explanation! And thanks as well to snibgo for the extra information on IM processing in general.

**Comment by snibgo** [comment]
@GeeMack:
> Since that "bogus" image has no filename suffix IM can't write it because it doesn't know how.

That is ingenious, maybe even genius. It isn't _just_ because the filename has no suffix, but _also_ because the input "xc:" has no format.

**Reply by GeeMack** [reply]
Thanks for the explanation. First I considered writing the placeholder image to a temp file, same file name every time, and overwriting it each time the command output the "bogus" image. That would require writing to disk for every image and deleting that single temporary file after the command. Pretty simple clean-up, but then I realized it didn't have to write the "bogus" image at all.
